src/http/server.cr

```diff
@@ -95,72 +100,187 @@ class HTTP::Server
+  def bind(port : Int32, reuse_port : Bool = false) : TCPServer
+    bind "127.0.0.1", port, reuse_port
+  end
+
+  # Creates a `TCPServer` listening on an unused port and adds it as an interface.
+  #
+  # Returns the determined port number.
+  #
+  # ```
+  # server = HTTP::Server.new { }
+  # server.bind_unused_port # => 12345
+  # ```
+  def bind_unused_port(host : String = "127.0.0.1", reuse_port : Bool = false) : Int32
+    tcp_server = bind(host, 0, reuse_port)
+    tcp_server.local_address.port
+  end
+
+  # Creates a `Socket::Server` listenting to *address* and adds it as an interface.
+  #
+  # The *address* can either be `host:port` or a Unix socket prefixed by `unix:`.
+  # This obviously works only on systems supporting Unix sockets.
   #
-  # For example you may let the system choose a port, then report it:
   # ```
-  # server = HTTP::Server.new(0) { }
-  # server.bind
-  # server.port # => 12345
+  # server = HTTP::Server.new { }
+  # server.bind "127.0.0.1:8080"
+  # server.bind "unix:/tmp/my-socket"
   # ```
-  def port
-    if server = @server
-      server.local_address.port.to_i
+  def bind(address : String, reuse_port : Bool = false) : Socket::Server
+    if address.starts_with?("unix:")
+      {% if flag?(:unix) %}
+      bind(Socket::UNIXAddress.new(address[5..-1])).as(Socket::Server)
+      {% else %}
+      raise NotImplementedError.new "system does not support Unix sockets"
+      {% end %}
     else
-      @port
+      host, _, port = address.partition(':')
+      bind(host, port.to_i, reuse_port).as(Socket::Server)
     end
   end
 
-  # Creates the underlying `TCPServer` if the doesn't already exist.
+  {% if flag?(:unix) %}
+  # Creates a `UNIXServer` bound to *address* and adds it as an interface.
   #
-  # You may set *reuse_port* to true to enable the `SO_REUSEPORT` socket option,
-  # which allows multiple processes to bind to the same port.
-  def bind(reuse_port = false)
-    @server ||= TCPServer.new(@host, @port, reuse_port: reuse_port)
+  # ```
+  # server = HTTP::Server.new { }
+  # server.bind Socket::UNIXAddress.new("/tmp/my-socket")
+  # ```
+  def bind(address : Socket::UNIXAddress) : UNIXServer
+    bind UNIXServer.new(address.path)
   end
+  {% end %}
 
-  # Starts the server. Blocks until the server is closed.
+  # Creates a `TCPServer` bound to *address* and adds it as an interface.
   #
-  # See `#bind` for details on the *reuse_port* argument.
-  def listen(reuse_port = false)
-    server = bind(reuse_port)
+  # ```
+  # server = HTTP::Server.new { }
+  # server.bind Socket::IPAddress.new("127.0.0.1", 12345)
+  # ```
+  def bind(address : Socket::IPAddress) : TCPServer
+    bind address.host, address.port, reuse_port
+  end
+
+  # Adds a `Socket::Server` *interface* to this server.
+  def bind(interface)
+    @interfaces << interface
+
+    interface
+  end
+
+  # Binds to the specified arguments and starts the server.
+  #
+  # See `#bind` and `#listen` for details
```

Comment:
Is it really helpful? We're introducing and require explicit `#bind` but in the same time we mix its logic (bind an interface) with `#listen` (for incoming connections).

Reply:
It's somewhat nice but not really necessary, I think. I guess we don't need it and it convolutes the API.

I will remove it. I desired, it can be added later.